cmds/core/readlink/readlink.go

```diff
@@ -49,27 +49,40 @@ func readLink(file string) error {
 		path, err = filepath.EvalSymlinks(file)
 	}
 
-	if *nonewline {
+	delimiter := "\n"
+	if *noNewLine {
 		delimiter = ""
 	}
 
-	fmt.Printf("%s%s", path, delimiter)
+	fmt.Fprintf(stdout, "%s%s", path, delimiter)
 	return err
 }
 
-func main() {
-	flag.Parse()
-
-	var exitStatus int
+func run(stdout io.Writer, stderr io.Writer, args []string) error {
+	if len(args) == 0 {
+		if *verbose {
+			fmt.Fprintf(stderr, "missing operand")
+		}
+		return fmt.Errorf("missing operand")
+	}
 
-	for _, file := range flag.Args() {
-		if err := readLink(file); err != nil {
+	var runErr error
+	for _, file := range args {
+		err := readLink(stdout, file)
+		if err != nil {
 			if *verbose {
-				fmt.Fprintf(os.Stderr, "%v\n", err)
+				fmt.Fprintf(stderr, "%v\n", err)
 			}
-			exitStatus = 1
+			runErr = err
 		}
 	}
 
-	os.Exit(exitStatus)
+	return runErr
+}
+
+func main() {
+	flag.Parse()
+	if err := run(os.Stdin, os.Stderr, flag.Args()); err != nil {
+		os.Exit(1)
```

Comment:
if you were to call log.Fatalf here we'd get a helpful error message and then an os.Exit(1)

Reply:
I think we will change expected behaviour if we add fatal. If we path multiple files to to readlink it will print all the links but return 1 if any of them are not links or not exists, or some other error. We are printing errors only if we path -v to the command.

Fatal will return 1, but also print bunch of staff.